src/api/posts/src/routes/posts.js

```diff
@@ -60,14 +60,31 @@ posts.get('/', validatePostsQuery(), async (req, res, next) => {
     first: `/posts?per_page=${perPage}&page=${1}`,
     last: `/posts?per_page=${perPage}&page=${Math.floor(postsCount / perPage)}`,
   });
-  res.json(
-    ids
+  let data = {};
+  if (expand === 1) {
+    try {
+      data = await Promise.all(
+        ids.map(async (id) => {
+          // obtain the corresponding feed to populate the author's name to the return data.
+          const post = await Post.byId(id);
+          return {
+            id,
+            url: `${postsUrl}/${id}`,
+            author: post.feed.author,
+            title: post.title,
+            publishDate: post.published,
+          };
+        })
+      );
+    } catch (err) {
+      logger.error({ err }, 'Unable to get detail information of posts from Redis');
+      next(createError(503, 'Unable to connect to database'));
+    }
+  } else
+    data = ids
```

Comment:
```suggestion
const data = ids.map((id) => ({ id, url: `${postsUrl}/${id}` }));

  if (expand === 1) {
    try {
      const expandedData = await Promise.all(
        data.map(async ({id, url}) => {
          // obtain the corresponding feed to populate the author's name to the return data.
          const post = await Post.byId(id);
          return {
            id,
            url,
            author: post.feed.author,
            title: post.title,
            publishDate: post.published,
          };
        })
      );
      
      res.json(expandedData);
    } catch (err) {
      logger.error({ err }, 'Unable to get detail information of posts from Redis');
      next(createError(503, 'Unable to connect to database'));
    }
  }
  
   res.json(data);
```

Suggestion: I think you could simplify like this to not repeats some logic

Reply:
Thank you very much. This would reduce our code's repetition. However, after the change, ``ids.map((id) => ({ id, url: `${postsUrl}/${id}` }));`` will always be executed regardless the expand parameter value. This could potentially slow down our code and it would be better if we keep it as it was.

Reply:
Yes, but you will need this data in both case anyway, also we are not processing millions of data, which I believe it won't slow down performance.

Reply:
I tried your solution and the result shows that when `expand=1` the posts services will crash since ` res.json(expandedData);` already sent back a json to the client and `res.json(data);` will use this already-consumed res object to send back a response to the client again. The posts service log looks like this: 
```
2022-11-19 18:40:12 Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
2022-11-19 18:40:12     at new NodeError (node:internal/errors:387:5)
2022-11-19 18:40:12     at ServerResponse.setHeader (node:_http_outgoing:644:11)
2022-11-19 18:40:12     at ServerResponse.header (/app/node_modules/.pnpm/express@4.18.2/node_modules/express/lib/response.js:794:10)
2022-11-19 18:40:12     at ServerResponse.send (/app/node_modules/.pnpm/express@4.18.2/node_modules/express/lib/response.js:174:12)
2022-11-19 18:40:12     at ServerResponse.json (/app/node_modules/.pnpm/express@4.18.2/node_modules/express/lib/response.js:278:15)
2022-11-19 18:40:12     at /app/src/routes/posts.js:87:7
2022-11-19 18:40:12     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
2022-11-19 18:40:12   code: 'ERR_HTTP_HEADERS_SENT'
2022-11-19 18:40:12 }
```

Reply:
Have you try to put `return res.json(expendedDate)`

Reply:
I just tried `return res.json(expendedDate)` and it works now. Thank you for your suggestion.